portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java

```diff
@@ -0,0 +1,79 @@
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import java.io.IOException;
+import java.io.PrintWriter;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/shoutbox")
+public class LoginServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("text/html;");
+    PrintWriter out = response.getWriter();
+    out.println("<h1>Log In</h1>");
+
+    // Only logged-in users can see the form
+    UserService userService = UserServiceFactory.getUserService();
+    if (userService.isUserLoggedIn()) {
+      out.println("<p>Hello " + userService.getCurrentUser().getEmail() + "!</p>");
+      out.println("<p>Type a message and click submit:</p>");
+      out.println("<form method=\"POST\" action=\"/shoutbox\">");
+      out.println("<textarea name=\"text\"></textarea>");
+      out.println("<br/>");
+      out.println("<button>Submit</button>");
+      out.println("</form>");
+    } else {
+      String loginUrl = userService.createLoginURL("/shoutbox");
+      out.println("<p>Login <a href=\"" + loginUrl + "\">here</a>.</p>");
+    }
+
+    // Everybody can see the messages
+    out.println("<ul>");
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query = new Query("Message").addSort("timestamp", SortDirection.DESCENDING);
+    PreparedQuery results = datastore.prepare(query);
+    for (Entity entity : results.asIterable()) {
+      String text = (String) entity.getProperty("text");
+      String email = (String) entity.getProperty("email");
+      out.println("<li>" + email + ": " + text + "</li>");
```

Comment:
and to be clear, something like this is ok because there isn't really a whole lot you can do since it's an arbitrary number of items.